realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/SyncedRealmMigrationTests.kt

```diff
@@ -0,0 +1,286 @@
+                .build()
+
+        // Setup initial Realm schema (with missing fields)
+        val className = StringOnly::class.java.simpleName
+        DynamicRealm.getInstance(config).use { dynamicRealm ->
+            val schema = dynamicRealm.schema
+            dynamicRealm.executeTransaction {
+                schema.create(className) // Create empty class
+            }
+        }
+
+        // Open typed Realm, which will validate the schema
+        Realm.getInstance(config).use { realm ->
+            assertTrue(realm.schema[className]!!.hasField(StringOnly.FIELD_CHARS)) // Field has been added
+        }
+    }
+
+    // Check that the Realm can still be opened even if the ondisk schema has more fields than in the model class.
+    // The underlying field should not be deleted, just hidden.
+    @Test
+    fun missingFields_hiddenSilently() {
+        val config = configFactory.createSyncConfigurationBuilder(createTestUser(app))
+                .schema(StringOnly::class.java)
+                .build()
+
+        // Setup initial Realm schema (with too many fields)
+        val className = StringOnly::class.java.simpleName
+        DynamicRealm.getInstance(config).use { dynamicRealm ->
+            val schema = dynamicRealm.schema
+            dynamicRealm.executeTransaction {
+                schema.create(className)
+                        .addField(StringOnly.FIELD_CHARS, String::class.java)
+                        .addField("newField", String::class.java)
+                // A schema version has to be set otherwise Object Store will try to initialize the schema again and reach an
+                // error branch. That is not a real case.
+                dynamicRealm.version = 0
+            }
+        }
+
+        // Open typed Realm, which will validate the schema
+        Realm.getInstance(config).use { realm ->
+            val stringOnlySchema = realm.schema[className]!!
+            assertTrue(stringOnlySchema.hasField(StringOnly.FIELD_CHARS))
+            assertTrue(stringOnlySchema.hasField("newField"))
+            assertEquals(2, stringOnlySchema.fieldNames.size.toLong())
+        }
+    }
+
+    // Check that a Realm cannot be opened if it contain breaking schema changes, like changing a primary key
+    @Test
+    fun breakingSchemaChange_throws() {
+        val config = configFactory.createSyncConfigurationBuilder(createTestUser(app))
+                .schema(PrimaryKeyAsString::class.java)
+                .build()
+
+        // Setup initial Realm schema (with a different primary key)
+        val expectedObjectSchema = OsObjectSchemaInfo.Builder(PrimaryKeyAsString.CLASS_NAME, 2, 0)
+                .addPersistedProperty(PrimaryKeyAsString.FIELD_PRIMARY_KEY, RealmFieldType.STRING, false, true, false)
+                .addPersistedProperty(PrimaryKeyAsString.FIELD_ID, RealmFieldType.INTEGER, true, true, true)
+                .build()
+        val list: MutableList<OsObjectSchemaInfo> = ArrayList()
```

Comment:
Does this needs to be a `MutableList`? Any reason not to just use `val list = listOf(expectedObjectSchema)` ?

Reply:
No. Updated